File: simulator/src/contract.ts
/**
 * Validation against the published contract.
 *
 * The simulator is the reference client of `api/openapi.yaml`, so it must not be
 * able to demonstrate anything the contract does not allow. Every envelope is
 * checked here *before* it reaches the network: a violation aborts the run
 * instead of being answered with a `400` the simulator would then have to
 * explain.
 *
 * The approach mirrors `api/scripts/validate-examples.mjs`: parse the contract,
 * register every component schema under its canonical `$ref` and compile it with
 * Ajv 2020 (OpenAPI 3.1 schemas are JSON Schema 2020-12). No bundling step is
 * needed because `api/openapi.yaml` contains no external `$ref`.
 */
import { readFileSync } from 'node:fs'
import { dirname, join } from 'node:path'
import { fileURLToPath } from 'node:url'

import Ajv2020, { type ValidateFunction } from 'ajv/dist/2020.js'
import addFormats from 'ajv-formats'
import { parse as parseYaml } from 'yaml'

import type { EventEnvelope } from './types.js'

/** Relative path of the contract inside the repository. */
export const CONTRACT_RELATIVE_PATH = 'api/openapi.yaml'

/**
 * Walks up from this module until it finds the contract. Resolving upwards
 * rather than hard-coding `../../api` keeps the lookup correct no matter whether
 * the simulator runs from `src/`, from a build output or from a test runner's
 * working directory.
 */
export function locateContract(startDir: string = dirname(fileURLToPath(import.meta.url))): string {
  let dir = startDir
  for (;;) {
    const candidate = join(dir, CONTRACT_RELATIVE_PATH)
    try {
      readFileSync(candidate, 'utf8')
      return candidate
    } catch {
      const parent = dirname(dir)
      if (parent === dir) {
        throw new Error(
          `could not find ${CONTRACT_RELATIVE_PATH} in any parent directory of ${startDir}`,
        )
      }
      dir = parent
    }
  }
}

export class ContractViolationError extends Error {
  constructor(
    readonly eventType: string,
    readonly clientEventId: string,
    readonly violations: string[],
  ) {
    super(
      `event ${clientEventId} of type ${eventType} violates ${CONTRACT_RELATIVE_PATH}:\n` +
        violations.map((v) => `    ${v}`).join('\n'),
    )
    this.name = 'ContractViolationError'
  }
}

export interface Contract {
  /** `info.version` of the contract the simulator validated against. */
  readonly documentVersion: string
  /** The closed v0 event catalogue, in the order the contract lists it. */
  readonly eventTypes: readonly string[]
  /** Absolute path of the contract document that was loaded. */
  readonly path: string
  /** Throws {@link ContractViolationError} when the envelope is not ingestible. */
  assertIngestible(event: EventEnvelope): void
  /** Validates a response body against a named component schema. */
  validateResponse(schemaName: string, body: unknown): string[]
}

interface OpenApiDocument {
  info?: { version?: string }
  components?: { schemas?: Record<string, Record<string, unknown>> }
}

/** Loads and compiles `api/openapi.yaml`. */
export function loadContract(contractPath: string = locateContract()): Contract {
  const spec = parseYaml(readFileSync(contractPath, 'utf8')) as OpenApiDocument
  const schemas = spec.components?.schemas
  if (!schemas) {
    throw new Error(`no component schemas found in ${contractPath}`)
  }

  const ajv = new Ajv2020({ strict: false, allErrors: true })
  // Format assertions are advisory in JSON Schema 2020-12. ADR 0004 records that
  // leaving them off silently accepted `"clientEventId": "not-a-uuid"` in the
  // backend; the simulator turns them on for the same reason.
  addFormats(ajv)
  for (const [name, schema] of Object.entries(schemas)) {
    ajv.addSchema(schema, `#/components/schemas/${name}`)
  }

  const compile = (name: string): ValidateFunction => {
    const validate = ajv.getSchema(`#/components/schemas/${name}`)
    if (!validate) {
      throw new Error(`schema ${name} does not exist in ${contractPath}`)
    }
    return validate
  }

  const ingest = schemas.IngestEventRequest
  const mapping = (ingest?.discriminator as { mapping?: Record<string, string> } | undefined)
    ?.mapping
  if (!mapping) {
    throw new Error(`IngestEventRequest in ${contractPath} has no discriminator mapping`)
  }

  const validateUnion = compile('IngestEventRequest')
  const branchCache = new Map<string, ValidateFunction>()

  const eventTypes = (schemas.EventType?.enum as string[] | undefined) ?? []
  if (eventTypes.length === 0) {
    throw new Error(`EventType in ${contractPath} has no enumeration`)
  }

  const describe = (validate: ValidateFunction): string[] =>
    (validate.errors ?? []).map((error) => `${error.instancePath || '/'} ${error.message ?? ''}`)

  return {
    documentVersion: spec.info?.version ?? 'unknown',
    eventTypes,
    path: contractPath,

    assertIngestible(event: EventEnvelope): void {
      // Validating straight against the 20-branch `oneOf` reports the failures
      // of the 19 branches the event never meant — ADR 0004 measured ~80 errors
      // for a single bad field. The contract's own discriminator mapping selects
      // the branch first, so the message names the real problem; the union is
      // still the authoritative gate afterwards.
      const ref = mapping[event.type]
      if (ref) {
        let branch = branchCache.get(ref)
        if (!branch) {
          const compiled = ajv.getSchema(ref)
          if (!compiled) {
            throw new Error(`discriminator maps ${event.type} to ${ref}, which does not exist`)
          }
          branch = compiled
          branchCache.set(ref, branch)
        }
        // Ajv types its validators as type guards over `unknown`, which would
        // narrow `event` to `never` in the failure branch. The explicit boolean
        // keeps the envelope typed while the check stays the same.
        const branchOk: boolean = branch(event)
        if (!branchOk) {
          throw new ContractViolationError(event.type, event.clientEventId, describe(branch))
        }
      }

      const unionOk: boolean = validateUnion(event)
      if (!unionOk) {
        throw new ContractViolationError(event.type, event.clientEventId, [
          `type ${event.type} matches no branch of IngestEventRequest`,
          ...describe(validateUnion),
        ])
      }
    },

    validateResponse(schemaName: string, body: unknown): string[] {
      const validate = compile(schemaName)
      return validate(body) ? [] : describe(validate)
    },
  }
}

File: backend/internal/ingest/contract.go
// Package ingest implements the write path of the event contract: the
// POST /api/v1/events endpoint, contract validation, the project lifecycle
// rules, RFC 9457 error responses and the post-commit publish hook.
//
// The layering is deliberate. This package decides *whether* an event may be
// accepted; the store decides *how* it is appended and projected. Nothing here
// writes to a read model, and nothing in the store knows an HTTP status code.
package ingest

import (
	_ "embed"
	"fmt"

	"github.com/pb33f/libopenapi"
	"github.com/pb33f/libopenapi-validator/config"
	schemavalidation "github.com/pb33f/libopenapi-validator/schema_validation"
	"github.com/pb33f/libopenapi/datamodel/high/base"
	"golang.org/x/text/language"
	"golang.org/x/text/message"
)

// contractDocument is the event contract the backend validates against.
//
// api/openapi.yaml is the authority, but it lives outside this Go module and
// go:embed cannot reach above the module root. The file below is therefore a
// verbatim copy, and TestEmbeddedContractMatchesTheAuthority fails the build as
// soon as the two differ — the copy cannot drift silently.
//
//go:embed contract/openapi.yaml
var contractDocument []byte

// ingestRequestSchema is the discriminated union of the ingestion request body.
// Its discriminator mapping is the single source of the event catalogue.
const ingestRequestSchema = "IngestEventRequest"

// schemaVersionSchema carries the enum of accepted payload schema versions.
const schemaVersionSchema = "SchemaVersion"

// componentSchemaPrefix is the local $ref prefix used throughout the contract.
const componentSchemaPrefix = "#/components/schemas/"

// Contract is the compiled event contract.
//
// Everything it knows — which event types exist, which schema versions are
// accepted and what each event must look like — is read out of the contract
// document at startup. There is no hand-maintained mirror of the spec in Go
// that could fall behind it.
type Contract struct {
	// eventSchemas maps an event type to the concrete envelope schema the
	// contract's discriminator selects for it.
	eventSchemas map[string]*base.Schema
	// eventTypes lists the catalogue in contract order.
	eventTypes []string
	// schemaVersions lists the accepted values of the schemaVersion field.
	schemaVersions []string

	validator schemavalidation.SchemaValidator
	printer   *message.Printer
}

// LoadContract parses the embedded contract and compiles it for validation.
//
// It fails loudly: a contract that cannot be read, or that no longer exposes
// the discriminated union the ingestion endpoint is built on, must keep the
// backend from starting rather than degrade validation at runtime.
func LoadContract() (*Contract, error) {
	document, err := libopenapi.NewDocument(contractDocument)
	if err != nil {
		return nil, fmt.Errorf("ingest: parsing the event contract: %w", err)
	}
	model, modelErr := document.BuildV3Model()
	if modelErr != nil {
		return nil, fmt.Errorf("ingest: building the contract model: %w", modelErr)
	}
	if model.Model.Components == nil || model.Model.Components.Schemas == nil {
		return nil, fmt.Errorf("ingest: the event contract declares no component schemas")
	}
	schemas := model.Model.Components.Schemas

	request, ok := schemas.Get(ingestRequestSchema)
	if !ok {
		return nil, fmt.Errorf("ingest: the event contract has no %s schema", ingestRequestSchema)
	}
	discriminator := request.Schema().Discriminator
	if discriminator == nil || discriminator.Mapping == nil {
		return nil, fmt.Errorf("ingest: %s carries no discriminator mapping", ingestRequestSchema)
	}

	contract := &Contract{
		eventSchemas: make(map[string]*base.Schema),
		// Format assertions are off by default in JSON Schema 2020-12, which
		// would silently accept a clientEventId that is not a UUID and an
		// occurredAt that is not a timestamp. The contract means them as
		// constraints, so they are switched on.
		validator: schemavalidation.NewSchemaValidator(config.WithFormatAssertions()),
		printer:   message.NewPrinter(language.English),
	}

	for pair := discriminator.Mapping.First(); pair != nil; pair = pair.Next() {
		eventType, ref := pair.Key(), pair.Value()
		name, err := schemaNameFromRef(ref)
		if err != nil {
			return nil, fmt.Errorf("ingest: event type %q: %w", eventType, err)
		}
		schema, ok := schemas.Get(name)
		if !ok {
			return nil, fmt.Errorf("ingest: event type %q maps to unknown schema %q", eventType, name)
		}
		contract.eventSchemas[eventType] = schema.Schema()
		contract.eventTypes = append(contract.eventTypes, eventType)
	}
	if len(contract.eventTypes) == 0 {
		return nil, fmt.Errorf("ingest: the event catalogue is empty")
	}

	version, ok := schemas.Get(schemaVersionSchema)
	if !ok {
		return nil, fmt.Errorf("ingest: the event contract has no %s schema", schemaVersionSchema)
	}
	for _, value := range version.Schema().Enum {
		if value != nil {
			contract.schemaVersions = append(contract.schemaVersions, value.Value)
		}
	}
	if len(contract.schemaVersions) == 0 {
		return nil, fmt.Errorf("ingest: %s declares no accepted values", schemaVersionSchema)
	}

	return contract, nil
}

// EventTypes returns the closed catalogue in contract order.
func (c *Contract) EventTypes() []string {
	return append([]string(nil), c.eventTypes...)
}

// SchemaVersions returns the accepted payload schema versions.
func (c *Contract) SchemaVersions() []string {
	return append([]string(nil), c.schemaVersions...)
}

// KnowsEventType reports whether the type is part of the closed catalogue.
func (c *Contract) KnowsEventType(eventType string) bool {
	_, ok := c.eventSchemas[eventType]
	return ok
}

// AcceptsSchemaVersion reports whether the payload schema version is accepted.
func (c *Contract) AcceptsSchemaVersion(version string) bool {
	for _, accepted := range c.schemaVersions {
		if accepted == version {
			return true
		}
	}
	return false
}

// Validate checks one decoded request body against the envelope schema the
// contract selects for eventType and returns every violation it finds.
//
// The document is validated against that single branch rather than against the
// whole oneOf: a union of twenty branches reports the nineteen the client never
// meant as failures too, which would bury the real errors. The discriminator is
// what the contract itself prescribes for choosing the branch.
func (c *Contract) Validate(eventType string, document any) []FieldError {
	schema, ok := c.eventSchemas[eventType]
	if !ok {
		return []FieldError{{
			Field:   pointerOf(nil),
			Code:    CodeInvalidEnum,
			Message: fmt.Sprintf("event type %q is not part of the v0 catalogue", eventType),
		}}
	}

	valid, failures := c.validator.ValidateSchemaObject(schema, document)
	if valid {
		return nil
	}

	// Every reported failure carries the same underlying validation tree, so
	// the collected violations are de-duplicated rather than concatenated.
	collector := newFieldErrorCollector(c.printer)
	for _, failure := range failures {
		for _, schemaFailure := range failure.SchemaValidationErrors {
			if schemaFailure.OriginalJsonSchemaError != nil {
				collector.collect(schemaFailure.OriginalJsonSchemaError)
				continue
			}
			// A failure without the underlying error still has to surface;
			// falling through silently would turn it into a false accept.
			collector.add(pointerOf(schemaFailure.InstancePath), CodeInvalidField, schemaFailure.Reason)
		}
	}
	return collector.errors()
}

// schemaNameFromRef turns a local component reference into its schema name.
func schemaNameFromRef(ref string) (string, error) {
	if len(ref) <= len(componentSchemaPrefix) || ref[:len(componentSchemaPrefix)] != componentSchemaPrefix {
		return "", fmt.Errorf("reference %q is not a local component schema", ref)
	}
	return ref[len(componentSchemaPrefix):], nil
}
